### ISSUE TITLE
Emitting from another process in async way.

### ISSUE THREAD

**Issue by bokolob** [miguelgrinberg/python-socketio#624]
Hello.
Now, AsyncPubSubManager.emit just makes await self._publish, without return. So, it is impossible to emit events from standalone async workers (run inside a thread is possible, but it smells bad - mixing different paradigms with no profit). 

I mean impossible to  do ` await socketio.emit("my_event", {})`

Is there a working solution?

If not, I can try to make a pull request.

PS.
I have a simple TCP server that reads data from IoT devices in an async manner and emits incoming events to WebSocket clients.

**Comment by miguelgrinberg** [comment]
Sorry, I don't follow your train of thought. Why can't you use the async pub/sub manager to emit?

**Comment by bokolob** [comment]
Hello.
Sorry for my unclear message.

Yes, of course, it's possible to use async manager and I do. But, it's not obvious which this call is
`socketio.emit("my_event", {})`  sync or async. Especially if you get socketio variable from somewhere outside. In my case different modules use different managers. It can lead to hidden bugs and timeouts.  If you write explicitly
`await socketio.emit("my_event", {})` it would be clear and checked by interpreter.

**Comment by miguelgrinberg** [comment]
I don't think you are doing this correctly, because the emit call is a coroutine. Are we talking about this one? https://github.com/miguelgrinberg/python-socketio/blob/master/socketio/asyncio_pubsub_manager.py#L39-L40.

This is the one you should use, and it is a coroutine as one would expect. Maybe you need to show some code so that there is no confusion?

**Comment by bokolob** [comment]
Yes, I talked about this code. But `await socketio.emit("my_event", {})` it doesn't work without changes. I get error 'object NoneType can't be used in 'await' expression'.

Couple of return  operators are needed.

```
 ...
    message_queue = AsyncRedisManager(Config.REDIS_URL, channel="flask-socketio")
    socketio = SocketIO(client_manager=message_queue, logger=True)
    socketio.init_app(None)
  
    ....
   
    def handler(data, item):
         await self.socketio.emit('coordinates_updated', {'data': {'id': item}},
                                                     room="__device_" + str(item))

```

**Comment by miguelgrinberg** [comment]
There has to be a misunderstanding. Let's see if we can clear this out.

You are using Flask-SocketIO, which is not compatible with async code. In your server process your message queue has to be the standard Redis manager class, since you can't use the async client in Flask. I believe you already must have this working correctly, but just in case I wanted to clarify it.

Your auxiliary process can be an async process if you like, but if that's the case then you can't use Flask-SocketIO in it, for the reasons stated above. Instead, you have to use the python-socketio package directly as documented here: https://python-socketio.readthedocs.io/en/latest/server.html#emitting-from-external-processes.

**Comment by bokolob** [comment]
It won't help, because the problem is located in python-socketio package.  So, using flask-socketio in my case could be incorrect, but it works (without await).

**Comment by miguelgrinberg** [comment]
@bokolob Okay, so trust me that I know what I'm talking about here. This is all code written by me, I'm not just giving you an opinion, I'm describing how I designed this to work.

I understand that this isn't working for you, and we are trying to figure out why. The solution that you are using is 100% incorrect. In an async application you cannot invoke a blocking function. At low scale it may appear that this works okay, but the blocking will not allow your application to scale properly.

If you don't care, then we can end this discussion and close the issue. If you want to figure out how to do this correctly, then humor me and implement your auxiliary process as I suggested above (while setting the channel name to `flask-socketio`, since python-socketio uses a different channel name by default).

Do this, and if it works, we celebrate. If it doesn't work, show me your code and we can continue discussing it until we figure it out 100%.
